### DISCUSSION TITLE
Allow `^` in arbitrary tasks

### DISCUSSION THREAD

**Discussion by mrmckeb** [vercel/turborepo#8999] [Ideas]
### Goals

I want to be able to use the following in `turbo.json` (note `"my-app#^build"`):

```json
{
  "extends": ["//"],
  "tasks": {
    "run": {
      "dependsOn": ["my-app#^build"]
    }
  }
}
```

That currently fails, only `"my-app#build"` is supported. see: https://turbo.build/repo/docs/reference/configuration#arbitrary-task-relationships

### Non-goals

No other changes are proposed.

### Background

Arbitrary dependencies are not the norm, but are sometimes required. This would be a helpful addition.

### Proposal

I'm not familiar with this part of the codebase, but would be happy to try.

**Comment by anthonyshew** [comment]
I'm confused about what that syntax would mean. Is it "Do `turbo run build --filter=my-app`, then `turbo run build`"?

**Reply by mrmckeb** [reply]
What we're looking for is a way to build the dependencies of a package, without building the package - does that help?

Sorry the original post was confusing!

**Reply by anthonyshew** [reply]
Ah, I see. This makes me think of two things:
- The [`...` microsyntax for `--filter`](https://turbo.build/repo/docs/reference/run#microsyntaxes-for-filtering) - but I can see how that would be insufficient if you want the task itself to *always* have this specifically defined behavior instead of with ad-hoc filters.
- With some creative script naming, you could do this:
```json
// turbo.json
{
  "tasks": {
    "build:app": {
      "dependsOn": ["^build"]
    }
  }
}
```

Then, your app would have:
```json
// apps/web/package.json
{
  "scripts": {
    "build:app": "the app's build script"
   }
}
```

And the libraries would have:
```json
// packages/ui/package.json
{
  "scripts": {
    "build": "tsc"
    }
  }
```

Writing that out, it _mostly_ strikes as a workaround without the feature you've proposed, but maybe that sounds sufficient to you? `turbo run build` would then only build dependent packages and `turbo run build:app` would build the app _and_ dependent packages. Curious to hear your thoughts, as I could see this ending up in an inefficient mess of tricky script naming.

**Reply by mrmckeb** [reply]
Thanks @anthonyshew, for now we've implemented a very similar pattern, but instead of renaming `build`, we've added a no-op script.

The app has a no-op script.
```jsonc
// apps/web/package.json
{
  "scripts": {
    "build": "next build",
    "build:deps": "exit 0;" // no-op
  }
}
```

That script has a dependency on dependencies being built.
```jsonc
// apps/web/turbo.json
{
  "tasks": {
    "build": {
      "dependsOn": ["^build"]
    },
    "build:deps": {
      "dependsOn": ["^build"]
    }
  }
}
```

We depend on the "build:deps" no-op script:
```jsonc
// pw/e2e/turbo.json
{
  "tasks": {
    "e2e": {
      "dependsOn": ["web#build:deps"]
    }
  }
}
```

This works, but it would be nice if we could use the `^` alongside the arbitrary dependency's script/task.

_(Side note: I just realised you can use `jsonc` instead of `json` in GitHub to allow comments not to show up as red errors!)_
